### PR TITLE
add blog post for canvas based renderer for matplotlib

### PR DESCRIPTION
**Pull request by madhur-tandon**
Adds a blog post for HTML5 canvas based renderer for the matplotlib library within pyodide
Context: https://github.com/pyodide/pyodide-blog/issues/3
CC: @rth 

Rendered version: https://deploy-preview-12--pyodide-blog.netlify.app/posts/canvas-renderer-matplotlib-in-pyodide/

*Estimated publication date*: 2022/1/25

### CODE REVIEW

**Comment by madhur-tandon** [comment]
One can access the deployed version of this PR here: https://deploy-preview-12--pyodide-blog.netlify.app/posts/canvas-renderer-matplotlib-in-pyodide/

**Comment by rth** [comment]
Thanks @madhur-tandon ! I'll provide feedback by the end of the week.

**Review by rth — COMMENTED** [review]
Thanks a lot for proposing it @madhur-tandon , it's an interesting read!

A few general suggestions are below, and more detailed comments are attached. Of course, feel to disagree with any of the suggestions :)

### Introduction

Because it's now going to be on the Pyodide blog, I think we don't need so much to present Pyodide (or if we do very briefly), however, it would be good to have a one or few sentences summary of what this post is about.  The title already does it, but maybe expand on it a bit.

### Motivation

Very clear.

### Challenges. Rendering Text:

It's fairly clear. Though I wonder if we could shorten the second part a bit by starting to say that the issue is that JS font loading is async, while draw_text in matplotlib is synchronous, talk about the two practical implications of that, and conclude on the solution chosen. 

### Rendering Images and Transparent Pixels:

I think this section could use some re-organization:
 - first start by mentioning the challenges being solved, rather than the way they were solved
 - also group text in more clearly defined paragraphs. Right now that section has 10 paragraphs, and it's difficult to follow how it's structured.

###  Results

This section could use a link to a demo say in jsfiddle. Also, IMO it should say that this new renderer was included in Pyodide in version ??, indicate how it can be enabled, and link to the (implementation)[https://github.com/pyodide/pyodide/blob/main/packages/matplotlib/src/html5_canvas_backend.py)

### How Fast is it?

Do you have the source for that benchmark? It would be good to re-run it. Various things were improved in Pyodide since.

### Potential Optimizations

Maybe there is no need to add a separate section for it, this could be merged with the previous section, particularly that both are relatively short.

File: content/posts/canvas-renderer-matplotlib-in-pyodide.md [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+---
+
+Pyodide compiles the Python Scientific Stack to WebAssembly (WASM) using Emscripten and was created by Michael Droettboom at Mozilla to be used within [Iodide](https://github.com/iodide-project/iodide). As of 2021, Pyodide now exists as an independent open source project.
+
+Bringing the scientific stack means adding support for the 5 most essential data science libraries --
+[numpy](https://numpy.org/), [pandas](https://pandas.pydata.org/), [matplotlib](https://matplotlib.org/), [scikit-learn](https://scikit-learn.org/), and [scipy](https://scipy.org/)
+
+### **Motivation**
+
+Supporting the above libraries requires packaging all of them, along with their dependencies. Unfortunately, browsers have limited memory, and if packages were to use it all, it would be hard to make some space for datasets.
```
[the remark this answers is not in the document]
Comment (hoodmane) [reply content/posts/canvas-renderer-matplotlib-in-pyodide.md:37]: Also, I think issue is less memory limits than the fact that it takes a long time to download and compile everything. We can quite comfortably load every single Pyodide package without running into memory limits.

File: content/posts/canvas-renderer-matplotlib-in-pyodide.md [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+
+However, as easy as it sounds, keeping track of when the font is loaded (so as to request a redraw) is not so trivial. The FontFace API is asynchronous and relies on the network -- making things harder due to the following problems:
+
+-  matplotlib can call *draw_text()* multiple times in a single drawing 
+    instance -- each of them requesting the *“correct”* font to be loaded. It’s 
+    possible that a previous request for dynamically loading the correct 
+    font isn’t completed yet (still fetching from the network), and a new 
+    request has been made at that time. This creates the problem of
+    multiple font loading due to queued and concurrent invocations of the      
+    FontFace API and affects performance.
```
[the remark this answers is not in the document]
Comment (hoodmane) [reply content/posts/canvas-renderer-matplotlib-in-pyodide.md:90]: I think it would probably be a better design to force people to preload the fonts they are going to use with an async call. It reminds me a bit of the Dolphin ubershaders:
https://dolphin-emu.org/blog/2017/07/30/ubershaders/

File: content/posts/canvas-renderer-matplotlib-in-pyodide.md [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+ShowPostNavLinks: true
+# cover:
+#     image: "<image path/url>" # image path/url
+#     alt: "<alt text>" # alt text
+#     caption: "<text>" # display caption under cover
+#     relative: false # when using page bundles set this to true
+#     hidden: true # only hide on current single page
+---
+
+Pyodide compiles the Python Scientific Stack to WebAssembly (WASM) using Emscripten and was created by Michael Droettboom at Mozilla to be used within [Iodide](https://github.com/iodide-project/iodide). As of 2021, Pyodide now exists as an independent open source project.
```
[the remark this answers is not in the document]
Comment (madhur-tandon) [reply content/posts/canvas-renderer-matplotlib-in-pyodide.md:30]: Thanks, modifying.

File: content/posts/canvas-renderer-matplotlib-in-pyodide.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (madhur-tandon) [reply content/posts/canvas-renderer-matplotlib-in-pyodide.md:37]: Thanks, noted and modifying.

File: content/posts/canvas-renderer-matplotlib-in-pyodide.md [hunk 4]
```diff
@@ -0,0 +1,169 @@
+---
+title: "HTML5 <canvas> based renderer for matplotlib in Pyodide"
```
[the remark this answers is not in the document]
Comment (madhur-tandon) [reply content/posts/canvas-renderer-matplotlib-in-pyodide.md:2]: Done

File: content/posts/canvas-renderer-matplotlib-in-pyodide.md [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+[numpy](https://numpy.org/), [pandas](https://pandas.pydata.org/), [matplotlib](https://matplotlib.org/), [scikit-learn](https://scikit-learn.org/), and [scipy](https://scipy.org/)
+
+### **Motivation**
+
+Supporting the above libraries requires packaging all of them, along with their dependencies. Unfortunately, browsers have limited memory, and if packages were to use it all, it would be hard to make some space for datasets.
+
+To make an effort in this direction -- we aimed to reduce the size of the matplotlib library.
+
+The [matplotlib](https://github.com/matplotlib/matplotlib) library has a component known as a renderer -- it's the renderer which defines the logic of how pixels get populated on the screen. The default renderer used by the matplotlib library is the [Agg](http://agg.sourceforge.net/antigrain.com/) renderer, which is a bunch of complicated C++ files. Currently, the version of matplotlib that is shipped with pyodide compiles the Agg renderer into WebAssembly. It’s the Agg renderer that draws the plot and an image of that plot is pasted (by extracting the underlying buffer data) onto the web document inside a canvas.
```
[the remark this answers is not in the document]
Comment (madhur-tandon) [reply content/posts/canvas-renderer-matplotlib-in-pyodide.md:41]: Done!

File: content/posts/canvas-renderer-matplotlib-in-pyodide.md [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+
+The Web already has some technologies to render graphics such as the *\<canvas\>* element and WebGL. Writing a new backend renderer based on these would enable us to reduce the size and memory footprint of the final build and possibly use GPU acceleration along with giving us the ability to use locally installed fonts with web fonts. In a Google Summer of Code project for 2019, we experimented with rendering graphics directly from matplotlib using the Canvas API.
```
[the remark this answers is not in the document]
Comment (madhur-tandon) [reply content/posts/canvas-renderer-matplotlib-in-pyodide.md:43]: Do you have some ideas regarding which points I could focus on more? Or should we write that this was also an experimental project to test the boundaries of what could be built by leveraging the `JavaScript <--> Python` interface?

File: content/posts/canvas-renderer-matplotlib-in-pyodide.md [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+```
+from js import document
+canvas_element = document.getElementById("canvas")
+ctx = canvas_element.getContext("2d")
+ctx.font = “30px cmr10”
+```
+
+The [cmr10](https://github.com/matplotlib/matplotlib/blob/main/lib/matplotlib/mpl-data/fonts/ttf/cmr10.ttf) font is available inside matplotlib’s virtual file system but is not accessible to the browser and thus, the *\<canvas\>* element cannot use it and falls back to using a default placeholder font.
+
+This problem is solved using [Web Fonts](https://developer.mozilla.org/en-US/docs/Web/API/FontFace) which allows us to dynamically load fonts with JavaScript and make it available to the browser. Once the *“correct”* font is loaded, we can request matplotlib to redraw the plot with the correct font. Again, all of this can be done inside Python since Pyodide can interact with browser APIs and other JavaScript libraries at a very fine level.
```
[the remark this answers is not in the document]
Comment (madhur-tandon) [reply content/posts/canvas-renderer-matplotlib-in-pyodide.md:80]: Done!

File: content/posts/canvas-renderer-matplotlib-in-pyodide.md [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+Thus, a simple *plt.plot()* invocation boils down to a *draw_path()* invocation for the backend which further calls the functions of the *\<canvas\>* element (responsible for actual rendering) -- all inside Python!
+
+### **Challenges**
+
+1. **Rendering Text**:
+
+The matplotlib library provides functionality for looking up fonts that it ships with. Unfortunately, these fonts are not accessible to the web browser to draw onto a canvas. We need to provide a way to load custom fonts that the browser can use.
+Thus, a simple assignment like the one below doesn’t work. 
+
+```
```
[the remark this answers is not in the document]
Comment (madhur-tandon) [reply content/posts/canvas-renderer-matplotlib-in-pyodide.md:71]: This should be ```python` actually. Did it.

File: content/posts/canvas-renderer-matplotlib-in-pyodide.md [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+    instance -- each of them requesting the *“correct”* font to be loaded. It’s 
+    possible that a previous request for dynamically loading the correct 
+    font isn’t completed yet (still fetching from the network), and a new 
+    request has been made at that time. This creates the problem of
+    multiple font loading due to queued and concurrent invocations of the      
+    FontFace API and affects performance.
+
+-  Drawing a plot from the matplotlib library leads to calling 
+    *“draw_text()”* which leads us to using the FontFace API to dynamically
+    load the font which leads us to requesting a redraw which further calls
```
[the remark this answers is not in the document]
Comment (madhur-tandon) [reply content/posts/canvas-renderer-matplotlib-in-pyodide.md:94]: Done, changed all instances of `dynamically` to `asynchronously ` as well.

File: content/posts/canvas-renderer-matplotlib-in-pyodide.md [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+    *“draw_text()”* and leads us to calling the FontFace API again which
+    then requests the redraw again...\
+    \
+    In short, it’s an infinite loop of\
```
[the remark this answers is not in the document]
Comment (madhur-tandon) [reply content/posts/canvas-renderer-matplotlib-in-pyodide.md:98]: Done

File: content/posts/canvas-renderer-matplotlib-in-pyodide.md [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+    `load font →  redraw →  load font →  redraw → ....`
+
+The above problems are similar but have a subtle difference amongst them. The
+first problem occurs for a single draw event (due to multiple *“draw_text”*
+invocations in that draw event), while the second problem leads to multiple
+never-ending redraws.
+
+Fortunately, both of these can be fixed using a global set which keeps track of
```
[the remark this answers is not in the document]
Comment (madhur-tandon) [reply content/posts/canvas-renderer-matplotlib-in-pyodide.md:106]: Done

File: content/posts/canvas-renderer-matplotlib-in-pyodide.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (madhur-tandon) [reply content/posts/canvas-renderer-matplotlib-in-pyodide.md:90]: I have replaced `dynamically` with `asynchronously`. This takes care of using the correct terms.

File: content/posts/canvas-renderer-matplotlib-in-pyodide.md [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+    *“draw_text()”* and leads us to calling the FontFace API again which
+    then requests the redraw again...\
+    \
+    In short, it’s an infinite loop of\
+    `load font →  redraw →  load font →  redraw → ....`
+
+The above problems are similar but have a subtle difference amongst them. The
+first problem occurs for a single draw event (due to multiple *“draw_text”*
+invocations in that draw event), while the second problem leads to multiple
+never-ending redraws.
```
[the remark this answers is not in the document]
Comment (madhur-tandon) [reply content/posts/canvas-renderer-matplotlib-in-pyodide.md:104]: Mentioned!

File: content/posts/canvas-renderer-matplotlib-in-pyodide.md [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+
+To overcome the above slowdown due to the rendering loop in Python, one solution is to rewrite this time-critical piece in JavaScript. This can be possible using the Function construct in JavaScript. The Function construct would essentially allow us to create JavaScript functions inside a python module capable of accepting and working with pythonic arguments.
+
+![snippet](https://user-images.githubusercontent.com/20173739/145672072-ef2f6bd2-7f37-436a-bbac-a324ccecc4cc.png)
+
+The above snippet defines a JavaScript function *_path_helper* which accepts Pythonic arguments and also supports calling of pythonic object’s methods inside it -- *path.iter_segments()*. The work related to this was attempted here: https://github.com/pyodide/pyodide/pull/510
+
+### **Conclusion**
+
+It’s been really gratifying to see all of the cool things that have been created with Pyodide in the short time since its public launch.  However, there’s still lots to do to turn this experimental proof-of-concept into a professional tool for everyday data science work. If you’re interested in helping us build that future, come find us on gitter, github and our mailing list.
```
[the remark this answers is not in the document]
Comment (madhur-tandon) [reply content/posts/canvas-renderer-matplotlib-in-pyodide.md:165]: Done

File: content/posts/canvas-renderer-matplotlib-in-pyodide.md [hunk 14]
```diff
@@ -0,0 +1,169 @@
...
+
+- The function *putImageData()* is not state-aware i.e. the image-data which is pasted doesn’t care about transformations. Thus, even if the canvas is transformed using operations such as rotate(), scale(), translate(), setTransform(). The resulting image would appear the same as if the canvas was not transformed at all.
+
+- The function *putImageData()* doesn’t care what is beneath it. If for some reason, the ImageData object contains some transparent pixels, and the canvas already has a red background, the transparent pixels will replace the red pixels which is not what we usually expect. We usually expect transparent pixels to blend with the background i.e. the pixels remain red at positions where pixels of the incoming ImageData object are transparent. But, this is not what happens, the previous pixels are simply over-written.
+
+### **Results**
+
+With the above major issues fixed, the new backend is ready. Here are some sample plots (taken from the matplotlib gallery) that are rendered using the new canvas based backend.
+
+![results](https://user-images.githubusercontent.com/20173739/144758542-c2a60150-dfb4-467e-96d0-db0403e68ec3.png)
```
[the remark this answers is not in the document]
Comment (madhur-tandon) [reply content/posts/canvas-renderer-matplotlib-in-pyodide.md:145]: Done

File: content/posts/canvas-renderer-matplotlib-in-pyodide.md [hunk 15]
```diff
@@ -0,0 +1,169 @@
...
+There are two functions in the *\<canvas\>* element which enable us to render images. 
+These are *putImageData()* and *drawImage()*.
+
+Images are essentially represented as multi-dimensional arrays and can be manipulated
+by the Numpy library on the Python side. The *\<canvas\>* element however needs the [ImageData](https://developer.mozilla.org/en-US/docs/Web/API/ImageData) object.
+
+Pyodide enables conversion of data types to and fro from JavaScript. Specifically, a
+bytes object in Python can be implicitly converted to a [Uint8ClampedArray](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Uint8ClampedArray) in JavaScript which is exactly what we require to create an ImageData object.
+
+Thus, a conversion from Numpy array to a raw bytes representation when interpreted as a Uint8ClampedArray in JavaScript can be used to render the ImageData object.
```
[the remark this answers is not in the document]
Comment (madhur-tandon) [reply content/posts/canvas-renderer-matplotlib-in-pyodide.md:123]: I have re-written the section so that challenges are written first. I have also included the snippet that does conversion from numpy to ImageData

File: content/posts/canvas-renderer-matplotlib-in-pyodide.md [hunk 16]
```diff
@@ -0,0 +1,169 @@
...
+
+![metrics](https://user-images.githubusercontent.com/20173739/62864633-e4b6ba00-bd29-11e9-8d98-92d72f339f68.png)
+
+In essence, the new renderer is about 2 to 2.5x slower but that’s the price one pays for reducing the size.
+
+### **Potential Optimizations**
+
+To overcome the above slowdown due to the rendering loop in Python, one solution is to rewrite this time-critical piece in JavaScript. This can be possible using the Function construct in JavaScript. The Function construct would essentially allow us to create JavaScript functions inside a python module capable of accepting and working with pythonic arguments.
+
+![snippet](https://user-images.githubusercontent.com/20173739/145672072-ef2f6bd2-7f37-436a-bbac-a324ccecc4cc.png)
```
[the remark this answers is not in the document]
Comment (madhur-tandon) [reply content/posts/canvas-renderer-matplotlib-in-pyodide.md:159]: The code isn't really complete. It has `...` for the if/else blocks aka not all if/else blocks are present. This was done to give a general idea and not really give the whole snippet. I have included the link to the closed PR which had the complete snippet for interested readers. Should I still include it here?

Also, the function was deliberately defined in Python to showcase that even Function API from JS can be used inside Python i.e. to boast off the power of how both Python and JS can interact with each other at a very fine level.

File: content/posts/canvas-renderer-matplotlib-in-pyodide.md [hunk 17]
```diff
@@ -0,0 +1,169 @@
...
+by the Numpy library on the Python side. The *\<canvas\>* element however needs the [ImageData](https://developer.mozilla.org/en-US/docs/Web/API/ImageData) object.
+
+Pyodide enables conversion of data types to and fro from JavaScript. Specifically, a
+bytes object in Python can be implicitly converted to a [Uint8ClampedArray](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Uint8ClampedArray) in JavaScript which is exactly what we require to create an ImageData object.
+
+Thus, a conversion from Numpy array to a raw bytes representation when interpreted as a Uint8ClampedArray in JavaScript can be used to render the ImageData object.
+
+The functions *putImageData()* and *drawImage()* can now be used. Both have different use-cases, but for our purpose we use them in tandem with something known as an In-memory canvas.
+
+An In-memory canvas is a normal *\<canvas\>* element which is used for off-screen rendering which is about rendering content somewhere, but the screen. That *“somewhere”* means the memory. Thus, we render graphics to the memory.
```
[the remark this answers is not in the document]
Comment (madhur-tandon) [reply content/posts/canvas-renderer-matplotlib-in-pyodide.md:127]: Explained before now and rewrote the section.

File: content/posts/canvas-renderer-matplotlib-in-pyodide.md [hunk 18]
```diff
@@ -0,0 +1,169 @@
...
+- The function *putImageData()* is not state-aware i.e. the image-data which is pasted doesn’t care about transformations. Thus, even if the canvas is transformed using operations such as rotate(), scale(), translate(), setTransform(). The resulting image would appear the same as if the canvas was not transformed at all.
+
+- The function *putImageData()* doesn’t care what is beneath it. If for some reason, the ImageData object contains some transparent pixels, and the canvas already has a red background, the transparent pixels will replace the red pixels which is not what we usually expect. We usually expect transparent pixels to blend with the background i.e. the pixels remain red at positions where pixels of the incoming ImageData object are transparent. But, this is not what happens, the previous pixels are simply over-written.
+
+### **Results**
+
+With the above major issues fixed, the new backend is ready. Here are some sample plots (taken from the matplotlib gallery) that are rendered using the new canvas based backend.
+
+![results](https://user-images.githubusercontent.com/20173739/144758542-c2a60150-dfb4-467e-96d0-db0403e68ec3.png)
+
```
[the remark this answers is not in the document]
Comment (madhur-tandon) [reply content/posts/canvas-renderer-matplotlib-in-pyodide.md:172]: This is a TODO for now.

**Comment by madhur-tandon** [comment]
@rth I have incorporated many changes according to most of the feedback. For things that are still left / need some discussion, they are still left unresolved and I have commented on them.

Some things that are left are:
- **Results section** -- include which version of Pyodide has this renderer, how it can be enabled, jsfiddle demo, implementation link
- **How Fast is it and Potential Optimisations section** -- source for benchmark (I think I will have to look for it / rewrite it)

**Comment by rth** [comment]
Thanks for addressing comments @madhur-tandon ! I think the up-to-date benchmarks and some live examples would be ideal. For examples, I also need to figure out how to make jsdelivr integration work for https://github.com/pyodide/pyodide-blog/pull/13

**Comment by madhur-tandon** [comment]
> up-to-date benchmarks and some live examples would be ideal

I will try to come up with them this coming weekend now that new snappier version is released as well

**Comment by madhur-tandon** [comment]
I was trying to recalculate benchmarks but it fails for me -- opened an issue here: https://github.com/pyodide/pyodide/issues/2150

**Comment by madhur-tandon** [comment]
@rth I have added new benchmarks. There is a significant performance boost with the current version of Pyodide. Hopefully this suffices. Please let me know any other changes that you may require.

File: content/posts/canvas-renderer-matplotlib-in-pyodide.md (hunk 18 above)
[the remark this answers is not in the document]
Comment (madhur-tandon) [reply content/posts/canvas-renderer-matplotlib-in-pyodide.md:170]: Let me know how to proceed with this! CC: @rth

File: content/posts/canvas-renderer-matplotlib-in-pyodide.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (madhur-tandon) [reply content/posts/canvas-renderer-matplotlib-in-pyodide.md]: Awaiting your response on this @rth :)

File: content/posts/canvas-renderer-matplotlib-in-pyodide.md [hunk 19]
```diff
@@ -0,0 +1,193 @@
+---
+title: "HTML5 <canvas> based renderer for Matplotlib in Pyodide"
+date: 2021-12-05T14:20:37+05:30
+draft: true
+tags: ["announcement"]
```
Comment (rth) [inline content/posts/canvas-renderer-matplotlib-in-pyodide.md:5]:
```suggestion
tags: ["visualization", "matplotlib"]
```

File: content/posts/canvas-renderer-matplotlib-in-pyodide.md [hunk 20]
```diff
@@ -0,0 +1,193 @@
...
+![snippet](/posts/2/snippet.png)
+
+The above snippet defines a JavaScript function *_path_helper* which accepts Pythonic arguments and also supports calling of pythonic object’s methods inside it -- *path.iter_segments()*. The work related to this was attempted here: https://github.com/pyodide/pyodide/pull/510
+
+### **Conclusion**
+
+It’s been really gratifying to see all of the cool things that have been created with Pyodide in the few years since its public launch.  However, there’s still lots to do to turn this proof-of-concept into a professional tool for everyday data science work. If you’re interested in helping us build that future, come find us on Gitter, Github and our mailing list.
+
+---
+
```
Comment (rth) [inline content/posts/canvas-renderer-matplotlib-in-pyodide.md:192]: Do you want to add an "About the author" section, as was done in https://github.com/pyodide/pyodide-blog/blame/master/content/posts/react-in-python-with-pyodide.md#L435

**Review by rth — COMMENTED** [review]
Thanks for your patience @madhur-tandon ! 

A few more comments are below, otherwise LGTM.

File: content/posts/canvas-renderer-matplotlib-in-pyodide.md (hunk 19 above)
Comment (madhur-tandon) [reply content/posts/canvas-renderer-matplotlib-in-pyodide.md:5]: Done, will push

**Comment by madhur-tandon** [comment]
@rth For the demo for `jsfiddle`, which version of Pyodide is used? Is it from the `main` branch? Or perhaps a particular release is used? I ask since, for the demo to work, we would need to merge some fixes that I did to the renderer in https://github.com/pyodide/pyodide/pull/2151

Since without those fixes, the `renderer` in the current `main` branch is kinda broken.
I have rebased that PR with the current changes in `main`.
If the CI times out, we can split `benchmarks` to 2 separate CI jobs.

**Comment by madhur-tandon** [comment]
@rth @ryanking13 I need help with the `jsfiddle` example.
1. How does `jsfiddle` have access to `pyodide`? Do we have to enable it somehow? Or perhaps it will be available by default?
2. Which version of Pyodide will be available? Is it the master branch or perhaps a specific release? I want to make sure I have my changes for my benchmark PR available with me.
3. To write python code inside `demo.js` file, I will have to use `pyodide.runPython` stuff right?

**Comment by madhur-tandon** [comment]
@rth I have added an example for `jsfiddle` but I am not sure if `0.19.1` contains the fixes in my benchmarks PR. Also, I could appreciate a review if I am doing this `demo` thingie in a correct way. Please let me know. Once you approve, I will finally add `About the Author` section in the next commit as well along with how to use my renderer in the blog in the `results` section.

**Comment by ryanking13** [comment]
@madhur-tandon Sorry for the late reply, we are going to release v0.20.0 before long, which includes your matplotlib patch. When it is released, you will be able to use a stable URL (https://cdn.jsdelivr.net/pyodide/v0.20.0/full/pyodide.js) in your example.

@rth Do you think this article is ready to be published (after the example is added)? We've made him wait too long.

File: content/posts/canvas-renderer-matplotlib-in-pyodide.md (hunk 20 above)
Comment (madhur-tandon) [reply content/posts/canvas-renderer-matplotlib-in-pyodide.md:203]: added, thanks

File: demos/canvas-renderer-matplotlib/demo-1/demo.js [hunk 21]
```diff
@@ -0,0 +1,37 @@
+async function main() {
+  let pyodide = await loadPyodide({
+    indexURL: "https://cdn.jsdelivr.net/pyodide/v0.19.1/full/",
+  });
```
Comment (rth) [inline demos/canvas-renderer-matplotlib/demo-1/demo.js:4]:
```suggestion
  let pyodide = await loadPyodide();
```
it's no longer necessary with 0.20.0a1

File: demos/canvas-renderer-matplotlib/demo-1/demo.html [hunk 22]
```diff
@@ -0,0 +1 @@
+<script src="https://cdn.jsdelivr.net/pyodide/v0.19.1/full/pyodide.js"></script>
```
Comment (rth) [inline demos/canvas-renderer-matplotlib/demo-1/demo.html:1]:
```suggestion
<script src="https://cdn.jsdelivr.net/pyodide/v0.20.0a1/full/pyodide.js"></script>
```

I can confirm that this worked.

File: content/posts/canvas-renderer-matplotlib-in-pyodide.md [hunk 23]
```diff
@@ -0,0 +1,204 @@
+---
+title: "HTML5 <canvas> based renderer for Matplotlib in Pyodide"
+date: 2021-12-05T14:20:37+05:30
```
Comment (rth) [inline content/posts/canvas-renderer-matplotlib-in-pyodide.md:3]:
```suggestion
date: 2021-03-31T12:00:00+01:00
```

**Review by rth — APPROVED** [review]
Sorry, this took so long @madhur-tandon, and thanks for your patience.

> How does jsfiddle have access to pyodide? Do we have to enable it somehow? Or perhaps it will be available by default?
> 
> Which version of Pyodide will be available? Is it the master branch or perhaps a specific release?

It's just an HTML file so you can load arbitrary JS library such any released versions of Pyodide including 0.20.0a1 which should have your fix. I can confirm that the demo works with this version. Though on mouse hover (for all versions) in the console I get,
```
Error: This borrowed proxy was automatically destroyed at the end of a function call. Try using create_proxy or create_once_callable.
The object was of type \&quot;method\&quot; and had repr \&quot;&lt;bound method FigureCanvasWasm.onmousemove of &lt;matplotlib.backends.html5_canvas_backend.FigureCanvasHTMLCanvas object at 0xab2640&gt;&gt;\&quot;"
```
(thanks @hoodmane  for improved error messages!)

so it sounds like this would need a fix in the MPL canvas backend. Though I think it's not critical for publishing this blog.

I reread the blog and everything looks good to me. 

I think we can update the release date and merge tomorrow, unless you wanted to make other changes?

**Comment by rth** [comment]
BTW, we will be only sure the link to jsfiddle works as expected, once we merge this PR.

File: content/posts/canvas-renderer-matplotlib-in-pyodide.md (hunk 23 above)
Comment (madhur-tandon) [reply content/posts/canvas-renderer-matplotlib-in-pyodide.md]: done, thanks

File: demos/canvas-renderer-matplotlib/demo-1/demo.html (hunk 22 above)
Comment (madhur-tandon) [reply demos/canvas-renderer-matplotlib/demo-1/demo.html]: done, thanks

File: demos/canvas-renderer-matplotlib/demo-1/demo.js (hunk 21 above)
Comment (madhur-tandon) [reply demos/canvas-renderer-matplotlib/demo-1/demo.js]: done, thanks

**Comment by rth** [comment]
This post is now [published](https://blog.pyodide.org/posts/canvas-renderer-matplotlib-in-pyodide/). Thanks again @madhur-tandon !

**Comment by hoodmane** [comment]
Thanks for your hard work and patience with this @madhur-tandon!
